polkadot/node/core/pvf/prepare-worker/src/lib.rs

```diff
@@ -328,3 +215,176 @@ fn runtime_construction_check(
+
+			// If we are pre-checking, check for runtime construction errors.
+			//
+			// As pre-checking is more strict than just preparation in terms of memory
+			// and time, it is okay to do extra checks here. This takes negligible time
+			// anyway.
+			if let PrepareJobKind::Prechecking = prepare_job_kind {
+				result = result.and_then(|output| {
+					runtime_construction_check(&output.0, &executor_params)?;
+					Ok(output)
+				});
+			}
+			result
+		},
+		Arc::clone(&condvar),
+		WaitOutcome::Finished,
+	)
+	.unwrap_or_else(|_| process::exit(libc::EXIT_FAILURE));
+
+	let outcome = thread::wait_for_threads(condvar);
+
+	match outcome {
+		WaitOutcome::Finished => {
+			let result =
+				prepare_thread.join().unwrap_or_else(|_| process::exit(libc::EXIT_FAILURE));
+			match result {
+				Ok(ok) => {
+					cfg_if::cfg_if! {
+						if #[cfg(target_os = "linux")] {
+							let (artifact, max_rss) = ok;
+						} else {
+							let artifact = ok;
+						}
+					}
+
+					// Stop the memory stats worker and get its observed memory stats.
+					#[cfg(any(target_os = "linux", feature = "jemalloc-allocator"))]
+					let memory_tracker_stats =
+						get_memory_tracker_loop_stats(memory_tracker_thread, process::id()).await;
+
+					let memory_stats = MemoryStats {
+						#[cfg(any(target_os = "linux", feature = "jemalloc-allocator"))]
+						memory_tracker_stats,
+						#[cfg(target_os = "linux")]
+						max_rss: extract_max_rss_stat(max_rss, process::id()),
+					};
+
+					let response: Result<Response, PrepareError> =
+						Ok(Response { artifact, memory_stats });
+
+					pipe_write
+						.write_all(response.encode().as_slice())
+						.unwrap_or_else(|_| process::exit(libc::EXIT_FAILURE));
+				},
+				Err(err) => {
+					pipe_write
+						.write_all(Err::<Response, PrepareError>(err).encode().as_slice())
+						.unwrap_or_else(|_| process::exit(libc::EXIT_FAILURE));
+				},
+			}
+
+			process::exit(libc::EXIT_SUCCESS);
+		},
+		_ => process::exit(libc::EXIT_FAILURE),
+	}
+}
+
+async fn handle_parent_process(
+	mut pipe_read: os_pipe::PipeReader,
+	temp_artifact_dest: PathBuf,
+	worker_pid: u32,
+	usage_before: Usage,
+	timeout: u64,
+) -> Result<PrepareStats, PrepareError> {
+	let mut received_data = Vec::new();
+
+	pipe_read
+		.read_to_end(&mut received_data)
+		.map_err(|err| PrepareError::Panic(err.to_string()))?;
+	let status = nix::sys::wait::wait();
+	let usage_after = nix::sys::resource::getrusage(UsageWho::RUSAGE_CHILDREN)
+		.map_err(|err| PrepareError::Panic(err.to_string()))?;
+	let cpu_tv = (get_total_cpu_usage(usage_after) - get_total_cpu_usage(usage_before)) as u64;
+
+	return match status {
+		Ok(nix::sys::wait::WaitStatus::Exited(_, libc::EXIT_SUCCESS)) => {
+			let result: Result<Response, PrepareError> = parity_scale_codec::decode_from_bytes(
+				bytes::Bytes::copy_from_slice(received_data.as_slice()),
+			)
+			.map_err(|e| PrepareError::Panic(e.to_string()))?;
```

Comment:
```suggestion
			let result: Result<Response, PrepareError> =
				Result::decode(&mut received_data.as_slice())
					.map_err(|e| PrepareError::Panic(e.to_string()))?;
```

Then we don't need the `bytes` dependency. :)